docs/src/tutorials/getting_started/design_patterns_for_larger_models.jl

```diff
@@ -0,0 +1,645 @@
+    optimize!(model)
+    return value.(x)
+end
+
+solve_knapsack_2(data)
+
+# ## Read in data from files
+
+# Having a data structure is a good step. But it is still annoying that we have
+# to hard-code the data into Julia. A good next step is to separate the data
+# into an external file format; JSON is a common choice.
+
+# The JuMP repository [has a file](https://github.com/jump-dev/JuMP.jl/blob/master/docs/src/tutorials/getting_started/data)
+# we're going to use for this tutorial. It is found here:
+
+data_filename = joinpath(@__DIR__, "data", "knapsack.json")
+
+# with these contents:
+
+println(read(data_filename, String))
+
+# Now let's write a function that reads this file and builds a `KnapsackData`
+# object:
+
+import JSON
+
+function read_data(filename)
+    d = JSON.parsefile(filename)
+    return KnapsackData(
+        Dict(
+            k => KnapsackObject(v["profit"], v["weight"]) for
+            (k, v) in d["objects"]
+        ),
+        d["capacity"],
+    )
+end
+
+data = read_data(data_filename)
+
+# ## Add options via if-else
+
+# At this point, we have data in a file format which we can load and solve a
+# single problem. For many users, this might be sufficient. However, at some
+# point you may be asked to add features like "but what if I want to take more
+# than one of a particular item?"
+
+# If this is the first time that you've been asked to add a feature, adding
+# options via `if-else` statements is a good approach. For example, we might
+# write:
+
+function solve_knapsack_3(data::KnapsackData; binary_knapsack::Bool)
+    model = Model(GLPK.Optimizer)
+    if binary_knapsack
+        @variable(model, x[keys(data.objects)], Bin)
+    else
+        @variable(model, x[keys(data.objects)] >= 0, Int)
+    end
+    @objective(model, Max, sum(v.profit * x[k] for (k, v) in data.objects))
+    @constraint(
+        model,
+        sum(v.weight * x[k] for (k, v) in data.objects) <= data.capacity,
+    )
+    optimize!(model)
+    return value.(x)
+end
+
+# Now we can solve the binary knapsack:
+
+solve_knapsack_3(data; binary_knapsack = true)
+
+# And an integer knapsack where we can take more than one copy of each item:
+
+solve_knapsack_3(data; binary_knapsack = false)
+
+# ## Add configuation options via dispatch
+
+# If you get repeated requests to add different options, you'll quickly find
+# yourself in a mess of different flags and `if-else` statements. It's hard to
+# write, hard to read, and hard to ensure you haven't introduced any bugs.
+# A good solution is to use Julia's type dispatch to control the configuration
+# of the model. The easiest way to explain this is by example.
+
+# First, start by defining a new abstract type, as well as new subtypes for each
+# of our options. These types are going to control the configuration of the
+# knapsack model.
+
+abstract type AbstractConfiguration end
+
+struct BinaryKnapsack <: AbstractConfiguration end
+
+struct IntegerKnapsack <: AbstractConfiguration end
```

Comment:
Would it make sense to rename the configs into "BinaryKnapsackProblem", "IntegerKnapsackProblem", "UpperBoundedKnapsackProblem"? I feel this makes it clearer that this config completely changes the type of problem.

Reply:
What about `BinaryConfig`, `IntegerConfig`, and UpperBoundedConfig`?

We're not really dispatching on the problem type. Dispatch is just a tool to configure the model in different ways. (e.g., you could use a config to set different solver arguments, or change the printing of names).

Reply:
I agree. In this case I'd personally favor `BinaryKnapsackConfig`, `IntegerKnapsackConfig` and, `UpperBoundedKnapsackConfig` but your suggestion also does the job :)